network/transport/v2/handlers.go

```diff
@@ -47,32 +48,66 @@ func (p protocol) Handle(peer transport.Peer, raw interface{}) error {
 	return nil
 }
 
-func (p protocol) handle(peer transport.Peer, envelope *Envelope) error {
-	switch msg := envelope.Message.(type) {
+type handleFunc func(peer transport.Peer, envelope *Envelope) error
+
+func (p *protocol) handleASynchWithLock(peer transport.Peer, envelope *Envelope, f handleFunc) error {
+	go func() {
+		if !p.mLock.TryRLock() {
+			log.Logger().Debugf("Sink active, ignoring %T (peer=%s)", envelope.Message, peer)
+			return
+		}
```

Comment:
could, but the `Unlock` remains within the go routine. That places the lock and unlock at different places. If the go routine is not scheduled for processing, the lock would remain. This would increase the chance for a long running lock.